apps/api/src/domain/services/media-item-query.service.ts

```diff
@@ -0,0 +1,89 @@
+import { Injectable } from '@nestjs/common';
+import mixTagsIntoViewModel from '../../app/controllers/utilities/mixTagsIntoViewModel';
+import { InternalError, isInternalError } from '../../lib/errors/InternalError';
+import { Maybe } from '../../lib/types/maybe';
+import { isNotFound, NotFound } from '../../lib/types/not-found';
+import cloneToPlainObject from '../../lib/utilities/cloneToPlainObject';
+import { RepositoryProvider } from '../../persistence/repositories/repository.provider';
+import { MediaItemViewModel } from '../../view-models/buildViewModelForResource/viewModels/media-item.view-model';
+import { MediaItem } from '../models/media-item/entities/media-item.entity';
+import { Tag } from '../models/tag/tag.entity';
+import IsPublished from '../repositories/specifications/isPublished';
+import { isResourceId } from '../types/ResourceId';
+import { resourceTypes } from '../types/resourceTypes';
+
+@Injectable()
+export class MediaItemQueryService {
+    constructor(private readonly repositoryProvider: RepositoryProvider) {}
+
+    async fetchById(id: unknown): Promise<InternalError | Maybe<MediaItemViewModel>> {
+        if (!isResourceId(id)) return new InternalError(`Invalid entity ID: ${id}`);
+
+        const searchResult = await this.repositoryProvider
+            .forResource<MediaItem>(resourceTypes.mediaItem)
+            .fetchById(id);
+
+        if (isInternalError(searchResult)) {
+            // One of the DTOs failed invariant validation
+            throw searchResult;
+        }
+
+        if (isNotFound(searchResult)) return NotFound;
+
+        const viewModel = new MediaItemViewModel(searchResult);
+
+        const viewModelWithTags = await this.mixinTheTags(viewModel);
+
+        return cloneToPlainObject(viewModelWithTags);
+    }
+
+    async fetchMany(): Promise<MediaItemViewModel[]> {
+        const searchResult = await this.repositoryProvider
+            .forResource<MediaItem>(resourceTypes.mediaItem)
+            .fetchMany(
+                // TODO support returning unpublished results for admin users
```

Comment:
It seem natural to take in a specification and leave it to the controller (adapter) to map user filters to `Specifications`, although we're not ready for that yet.